### PR TITLE
[llvm-objcopy] Add support of symbol modification flags for MachO

### LINKED ISSUE #120894: llvm-objcopy MachO: does not support localizing and globalizing symbols
On some occasions, there is a need to modify the behavior of a translation unit in a static library. There are many ways to achieve the same result, but the least hacky option might be exposing private symbols in an object file.
Currently, llvm-objcopy for Mach-O does not support either `--localize-symbol` or `--globalize-symbol`.

https://github.com/RIscRIpt/llvm-project/blob/b28412d5397dc5a23f172d6ebeac760c82a82248/llvm/lib/ObjCopy/ConfigManager.cpp#L36-L37

### PR DESCRIPTION
**Pull request by ridzenis**
This patch adds support of the following llvm-objcopy flags for MachO:

- `--globalize-symbol`, `--globalize-symbols`,
- `--keep-global-symbol`, `-G`, `--keep-global-symbols`,
- `--localize-symbol`, `-L`, `--localize-symbols`,
- `--skip-symbol`, `--skip-symbols`.

Code in `updateAndRemoveSymbols` for MachO
is kept similar to its version for ELF.

Fixes #120894

### CODE REVIEW

**Comment by RIscRIpt** [comment]
I was not sure if I have to adjust `N_PEXT` flag in case we are applying either "localization" or "globalization". So I just keep it unchanged. Please let me know if this is not correct.

**Comment by RIscRIpt** [comment]
Fixed tests on Windows, and rebased onto main. I did a force-push, because none has submitted review comments yet.

File: llvm/lib/ObjCopy/MachO/MachOObject.cpp
Comment (alexander-shaposhnikov) [inline llvm/lib/ObjCopy/MachO/MachOObject.cpp:41]: nit: https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly 
```it_ext -> ExternalSymbolsBegin```

File: llvm/lib/ObjCopy/MachO/MachOObjcopy.cpp
Comment (jh7370) [inline llvm/lib/ObjCopy/MachO/MachOObjcopy.cpp:101]: Nit: is the `llvm::` prefix here and in various other places needed? This code is already inside the `llvm` namespace.

File: llvm/test/tools/llvm-objcopy/MachO/globalize-symbol.test
Comment (jh7370) [inline llvm/test/tools/llvm-objcopy/MachO/globalize-symbol.test:5]: Nit: when we have a big block like this of CHECK followed by lots of CHECK-NEXT, we tend to add a few spaces after the colon before the actual CHECK, per the suggestion, to make the checked-for text line up with the remaining lines below. Applies throughout.
```suggestion
# CHECK:      Symbols [
```

**Review by jh7370 — COMMENTED** [review]
Please update the docs (llvm/docs/CommandGuide/llvm-objcopy.rst) to move the relevant options from the ELF-specific section to the generic options section.

Am I right in thinking that this automatically supports the file-based inputs like `--globalize-symbols` too, because this bit is in the generic code? If so, please update the tests to cover them and the PR description to mention this.

**Comment by RIscRIpt** [comment]
> Please update the docs (llvm/docs/CommandGuide/llvm-objcopy.rst) to move the relevant options from the ELF-specific section to the generic options section.

Done. I've also added a short notes to each option that it is supported only for ELF and MachO. Please check it.

Additionally, I've updated llvm/ReleaseNotes.md

> Am I right in thinking that this automatically supports the file-based inputs like --globalize-symbols too, because this bit is in the generic code? If so, please update the tests to cover them and the PR description to mention this.

That's right. I've updated PR description, and tests.

File: llvm/docs/CommandGuide/llvm-objcopy.rst
Comment (jh7370) [inline llvm/docs/CommandGuide/llvm-objcopy.rst:86]: FWIW, I don't think we have this note on other options that are specific to a subset of the supported formats we have, so I'd omit it to avoid confusion.

File: llvm/docs/ReleaseNotes.md
Comment (jh7370) [inline llvm/docs/ReleaseNotes.md:356]: I think you can omit `-G` and `-L` as they're just the short aliases for some of the other option names.

File: llvm/docs/ReleaseNotes.md
Comment (jh7370) [inline llvm/docs/ReleaseNotes.md:362]: Related to how cherry-picks work, I'd put this in its own separate commit that you push directly, since it's completely unrelated.

File: llvm/docs/ReleaseNotes.md
Comment (jh7370) [inline llvm/docs/ReleaseNotes.md:353]: I think technically the spelling is "Mach-O". We've not done a great job in the command guide of being consistent, so fixes there welcome too (in a separate commit).

**Review by jh7370 — APPROVED** [review]
LGTM from my point of view, but please wait for @alexander-shaposhnikov, as I don't know Mach-O really, so can't comment on the format-specific details.

File: llvm/docs/CommandGuide/llvm-objcopy.rst
Comment (RIscRIpt) [reply llvm/docs/CommandGuide/llvm-objcopy.rst]: Removed.

File: llvm/docs/ReleaseNotes.md
Comment (RIscRIpt) [reply llvm/docs/ReleaseNotes.md]: This was the only occurrence in ReleaseNotes, so I fixed it in a fixup commit.

File: llvm/docs/ReleaseNotes.md
Comment (RIscRIpt) [reply llvm/docs/ReleaseNotes.md]: Removed.

File: llvm/docs/ReleaseNotes.md
Comment (RIscRIpt) [reply llvm/docs/ReleaseNotes.md]: Reverted in a fixup. I'll push a separate commit to main directly.

**Comment by RIscRIpt** [comment]
Thank you for the review @jh7370

**Comment by alexander-shaposhnikov** [comment]
LG

**Review by alexander-shaposhnikov — APPROVED** [review]

**Comment by RIscRIpt** [comment]
Thank you for reviews! I took another look at my tests and realized that I can remove `LC_DYSYMTAB` load command - that's what I did in the another fixup commit.

I'll merge this PR once all pre-merge checks pass.

**Pull request merged**
